scalameta/trees/shared/src/main/scala/scala/meta/Trees.scala

```diff
@@ -417,12 +417,12 @@ object Defn {
   @ast class Def(
       mods: List[Mod],
       name: Term.Name,
-      tparams: List[scala.meta.Type.Param],
-      paramss: List[List[Term.Param]],
+      paramss: List[Either[List[scala.meta.Type.Param],List[Term.Param]]],
```

Comment:
I think this is more likely to be
```
List[(List[Type.Param], @nonEmpty List[Term.Param])]
```
or
```
List[(Option[List[Type.Param]], @nonEmpty List[Term.Param])]
```

since it shouldn't allow `[A][B1, B2](a: A)`, which then doesn't require `Either` handling.

Reply:
It does allow the above !
And since `[A]: A` is allowed, `@nonEmpty` can't be used
So we really need something like `List[List | List]`, that's more or less how it is implemented in the candidate implementation for clause interleaving

Reply:
> It does allow the above ! And since `[A]: A` is allowed, `@nonEmpty` can't be used So we really need something like `List[List | List]`, that's more or less how it is implemented in the candidate implementation for clause interleaving

what is `[A]: A`? is that `def foo([...]: ...)`?

If `[...][...]` is not allowed (as it shouldn't, i think), either will not be structured enough.

Reply:
I meant `def foo[A]: A`, which is already valid

With clause interleaving, `def bar[A][B]` will be allowed

Therefore we really need some form of disjunct union, I chose `Either`, but it could also be done with two `case class`es for example

Reply:
i would forgo Either and define a new Tree which contains just tparams and params, so that you can capture positions, tokens etc.

however, you must also make changes to parsing of Term.Apply and ApplyType.